src/tomography.jl

```diff
@@ -0,0 +1,244 @@
+# n : number of qubits
+
+#Construct the predictor matrix.  Each row is an experiment.  The number of
+#columns is 4^n for the size of the vectorized density matrix.
+#First transform the measurement operators by the readout pulses to create
+#the effective measurement operators and then flatten into row of the
+#predictor matrix
+
+@printf("Setting up predictor matrix....");
+ predictorMat = complex(zeros(length(expResults), 4^n));for expct = 1:length(expResults)
+  tmp = transpose(measPulseUs[Int(measPulseMap[Int(expct)])]'*measOps[Int(measOpMap[Int(expct)])]*measPulseUs[Int(measPulseMap[Int(expct)])]);
+     predictorMat[expct,:] = tmp[:];
+end
+@printf("Done!\n.")
+
+@printf("SDP optimizing...")
+
+solver = SCSSolver(verbose=0, max_iters=10_000, eps = 1e-8)
+
+invVarMat = inv(varMat);
+invVarMat = real(sqrtm(real(sqrtm(invVarMat'*invVarMat))));
+
+
+ρr = Variable(2^n, 2^n)
+ρi = Variable(2^n, 2^n)
+
+constraints = trace(ρr) == 1
+constraints += trace(ρi) == 0
+constraints += isposdef([ρr ρi; -ρi ρr])
+
+#We want to minimize the difference between predicted results and experimental results
+problem = minimize( vecnorm(invVarMat*(expResults - [real(predictorMat) imag(predictorMat)]*[vec(ρr); vec(ρi)]), 2)^2, constraints )
+solve!(problem, solver)
+
+@printf("Done\n")
+return (ρr.value - 1im*ρi.value)
+end
+
+
+function analyzeStateTomo2(data, nbrQubits, nbrPulses, nbrCalRepeats)
+
+    numMeas = length(data);
+    measOps = Dict();
+    tomoData = Array(Float64[]);
+    varData = Array(Float64[]);
+    if haskey(data, "data")
+        datatemp = Dict()
+        datatemp[1]=data
+        numMeas=1
+    end
+for ct = 1:numMeas
+    #Average over calibration repeats
+    calData = mean(reshape(real(datatemp[ct]["data"][end-nbrCalRepeats*(2^nbrQubits)+1:end]), nbrCalRepeats, 2^nbrQubits), 1);
+    #Pull out the calibrations as diagonal measurement operators
+    measOps[ct] = diagm(calData[:]);
+
+    #The data to invert
+
+        tomoData = [tomoData;  real(datatemp[ct]["data"][1:end-nbrCalRepeats*(2^nbrQubits)])];
+
+    #variance
+        varData = [varData; datatemp[ct]["realvar"][1:end-nbrCalRepeats*(2^nbrQubits)]];
+
+end
+#Map each experiment to the appropriate readout pulse
+    measOpMap = reshape(repmat(transpose(1:numMeas), nbrPulses^nbrQubits, 1), numMeas*(nbrPulses^nbrQubits), 1);
+
+    measPulseMap = repmat((1:nbrPulses^nbrQubits), numMeas, 1);
+#Use a helper to get the measurement unitaries.
+measPulseUs = tomo_gate_set(nbrQubits, nbrPulses);
+varMat = diagm(varData[:]);
+
+#Now call the inversion routines
+
+#First least squares
+    rhoLSQ = QST_LSQ(tomoData, varMat, measPulseMap, measOpMap, measPulseUs, measOps, nbrQubits);
+#plotting to be implemented    pauliSetPlot(rho2pauli(rhoLSQ), newplot);
+
+#Now constrained SDP
+    rhoSDP = QST_SDP(tomoData, varMat, measPulseMap, measOpMap, measPulseUs, measOps, nbrQubits);
+
+    return (rhoLSQ, rhoSDP)
+end
+
+function rho2pauli(rho)
+#converts a density matrix to a Pauli set
```

Comment:
using Cliffords.jl we can write this a rather different way that would be much clearer. I think the central loop ought to look like:
```julia
paulis = [kron(x...) for x in product(repeated([Id, X, Y, Z], nbrQubits)...)]
paulis = sort(paulis)
# the above requires a definition of isless(a::Pauli, b::Pauli)... I will add this to Cliffords.jl
for (ct, p) in enumerate(paulis):
    pauliVec[ct] = real(trace(rho * p))
end
```